File: apps/mobile/test/widgets/report_button_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile/core/api_client.dart';
import 'package:mobile/models/report_reason.dart';
import 'package:mobile/widgets/report_button.dart';

void main() {
  testWidgets('un signalement réussi passe le bouton sur "Signalé"', (
    tester,
  ) async {
    ReportReason? reportedReason;

    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: ReportButton(
            onReport: (reason) async => reportedReason = reason,
          ),
        ),
      ),
    );

    await tester.tap(find.text('Signaler'));
    await tester.pumpAndSettle();
    await tester.tap(find.text('Confirmer'));
    await tester.pumpAndSettle();

    expect(reportedReason, ReportReason.spam);
    expect(find.text('Signalé'), findsOneWidget);
  });

  testWidgets('un 409 (déjà signalé) est traité comme un succès silencieux', (
    tester,
  ) async {
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: ReportButton(
            onReport: (reason) async =>
                throw const ApiException(409, 'Déjà signalé'),
          ),
        ),
      ),
    );

    await tester.tap(find.text('Signaler'));
    await tester.pumpAndSettle();
    await tester.tap(find.text('Confirmer'));
    await tester.pumpAndSettle();

    expect(find.text('Signalé'), findsOneWidget);
    expect(find.byType(SnackBar), findsNothing);
  });
}

File: apps/web/src/lib/components/ui/ReportButton.test.ts
import { describe, expect, it, vi } from 'vitest';
import { render, screen, fireEvent, waitFor } from '@testing-library/svelte';
import ReportButton from './ReportButton.svelte';
import { ApiError } from '$lib/api/client';

describe('ReportButton', () => {
	it('un signalement réussi passe le bouton sur "Signalé"', async () => {
		const onReport = vi.fn().mockResolvedValue(undefined);
		render(ReportButton, { onReport });

		await fireEvent.click(screen.getByRole('button', { name: 'Signaler' }));
		await fireEvent.click(screen.getByRole('button', { name: 'Confirmer' }));

		await waitFor(() => expect(onReport).toHaveBeenCalledWith('SPAM'));
		expect(screen.getByRole('button', { name: 'Signalé' })).toBeInTheDocument();
	});

	it('un 409 (déjà signalé) est traité comme un succès silencieux', async () => {
		const onReport = vi.fn().mockRejectedValue(new ApiError(409, 'Déjà signalé'));
		render(ReportButton, { onReport });

		await fireEvent.click(screen.getByRole('button', { name: 'Signaler' }));
		await fireEvent.click(screen.getByRole('button', { name: 'Confirmer' }));

		await waitFor(() =>
			expect(screen.getByRole('button', { name: 'Signalé' })).toBeInTheDocument()
		);
		expect(screen.queryByRole('alert')).not.toBeInTheDocument();
	});
});
